### PR TITLE
Theming: Support for runtime theme switching and hooks for custom themes

### PR DESCRIPTION
**Pull request by torkelo**
Initial changes needed for custom themes & runtime change of themes (See enterprise repo for paired PR)

### CODE REVIEW

**Review by marefr — APPROVED** [review]
Backend changes looks good :)

File: packages/jaeger-ui-components/src/TracePageHeader/SpanGraph/Scrubber.tsx [hunk 1]
```diff
@@ -19,50 +19,56 @@ import { createStyle } from '../../Theme';
...
       }
-      &:hover + .${ScrubberLine} {
+      &:hover + .scrubber.line {
         stroke: #44f;
       }
     `,
-    ScrubberHandleExpansion,
-    ScrubberHandle,
-    ScrubberLine,
   };
```
Comment (torkelo) [inline packages/jaeger-ui-components/src/TracePageHeader/SpanGraph/Scrubber.tsx:72]: These changes are not related to main change, just updated a PageHeader test and got some css emotion warnings that made reading the jest errors hard

File: packages/jaeger-ui-components/src/TracePageHeader/TracePageHeader.tsx [hunk 2]
```diff
@@ -163,13 +165,13 @@ export const HEADER_ITEMS = [
...
+    renderer(trace: Trace, styles: ReturnType<typeof getStyles>) {
       const dateStr = formatDatetime(trace.startTime);
       const match = dateStr.match(/^(.+)(:\d\d\.\d+)$/);
       return match ? (
-        <span className={styles?.TracePageHeaderOverviewItemValue}>
+        <span className={styles.TracePageHeaderOverviewItemValue}>
           {match[1]}
-          <span className={styles?.TracePageHeaderOverviewItemValueDetail}>{match[2]}</span>
+          <span className={styles.TracePageHeaderOverviewItemValueDetail}>{match[2]}</span>
         </span>
```
Comment (torkelo) [inline packages/jaeger-ui-components/src/TracePageHeader/TracePageHeader.tsx:175]: Not related to main change, just fixing emotion warnings

File: public/app/core/components/PageHeader/PageHeader.test.tsx [hunk 3]
```diff
@@ -1,12 +1,10 @@
 import React from 'react';
 import PageHeader from './PageHeader';
-import { shallow, ShallowWrapper } from 'enzyme';
+import { render } from '@testing-library/react';
```
Comment (Clarity-89) [inline public/app/core/components/PageHeader/PageHeader.test.tsx:3]:
```suggestion
import { render, screen } from '@testing-library/react';
```
Let's use `screen` to be consistent across tests.

File: public/app/core/components/PageHeader/PageHeader.test.tsx [hunk 4]
```diff
@@ -17,17 +15,16 @@ describe('PageHeader', () => {
         },
         node: {},
       };
-      wrapper = shallow(<PageHeader model={nav as any} />);
-    });
 
-    it('should render the title', () => {
-      const title = wrapper.find('.page-header__title');
-      expect(title.text()).toBe('node');
+      const dom = render(<PageHeader model={nav as any} />);
```
Comment (Clarity-89) [inline public/app/core/components/PageHeader/PageHeader.test.tsx:19]:
```suggestion
      render(<PageHeader model={nav as any} />);
```

File: public/app/core/components/PageHeader/PageHeader.test.tsx [hunk 5]
```diff
@@ -17,17 +15,16 @@ describe('PageHeader', () => {
...
+
+      const title = await dom.findByTestId('page-title');
```
Comment (Clarity-89) [inline public/app/core/components/PageHeader/PageHeader.test.tsx:21]:
```suggestion
      const title = await screen.findByTestId('page-title');
```

File: public/app/core/components/PageHeader/PageHeader.test.tsx [hunk 6]
```diff
@@ -39,15 +36,14 @@ describe('PageHeader', () => {
         },
         node: {},
       };
-      wrapper = shallow(<PageHeader model={nav as any} />);
-    });
 
-    it('should render the title with breadcrumbs first and then title last', () => {
-      const title = wrapper.find('.page-header__title');
-      expect(title.text()).toBe('Parent / child');
+      const dom = render(<PageHeader model={nav as any} />);
```
Comment (Clarity-89) [inline public/app/core/components/PageHeader/PageHeader.test.tsx:40]:
```suggestion
      render(<PageHeader model={nav as any} />);
```

File: public/app/core/components/PageHeader/PageHeader.test.tsx [hunk 7]
```diff
@@ -39,15 +36,14 @@ describe('PageHeader', () => {
...
+
+      const title = await dom.findByTestId('page-title');
```
Comment (Clarity-89) [inline public/app/core/components/PageHeader/PageHeader.test.tsx:42]:
```suggestion
      const title = await screen.findByTestId('page-title');
```

File: public/app/core/components/PageHeader/PageHeader.tsx [hunk 8]
```diff
@@ -71,86 +71,92 @@ const Navigation = ({ children }: { children: NavModelItem[] }) => {
...
-    breadcrumbsResult.push(<span key={breadcrumbs.length + 1}> / {title}</span>);
+function renderHeaderTitle(main: NavModelItem) {
+  const iconClassName =
+    main.icon === 'grafana'
+      ? css`
+          margin-top: 12px;
+        `
+      : css`
+          margin-top: 14px;
+        `;
```
Comment (Clarity-89) [inline public/app/core/components/PageHeader/PageHeader.tsx:104]:
```suggestion
   const iconClassName = css`
    margin-top: ${main.icon === 'grafana' ? '12px' : '14px'};
  `;
```

File: public/app/core/components/PageHeader/PageHeader.tsx [hunk 9]
```diff
@@ -71,86 +71,92 @@ const Navigation = ({ children }: { children: NavModelItem[] }) => {
...
-          {main.icon && <Icon name={main.icon as IconName} size="xxxl" className={iconClassName} />}
-          {main.img && <img className="page-header__img" src={main.img} alt={`logo of ${main.text}`} />}
-        </span>
-
-        <div className="page-header__info-block">
-          {this.renderTitle(main.text, main.breadcrumbs ?? [])}
-          {main.subTitle && <div className="page-header__sub-title">{main.subTitle}</div>}
-        </div>
-      </div>
+      <h1 className="page-header__title" data-testid="page-title">
```
Comment (Clarity-89) [inline public/app/core/components/PageHeader/PageHeader.tsx:128]: `data-testid` shouldn't be necessary here. You'd be able to query this element via `screen.getByRole('heading', {name: [the string that title evaluates to]})`.

File: public/app/core/components/PageHeader/PageHeader.tsx [hunk 10]
```diff
@@ -71,86 +71,92 @@ const Navigation = ({ children }: { children: NavModelItem[] }) => {
...
-  render() {
-    const { model } = this.props;
-
-    if (!model) {
-      return null;
+  const breadcrumbsResult = [];
+  for (const bc of breadcrumbs) {
+    if (bc.url) {
+      breadcrumbsResult.push(
+        <a className="text-link" key={breadcrumbsResult.length} href={bc.url} data-testid="breadcrumb-text-link">
```
Comment (Clarity-89) [inline public/app/core/components/PageHeader/PageHeader.tsx:138]: data-testid shouldn't be necessary here. You'd be able to query this element via `screen.getByRole('link', {name: [the string that title evaluates to]})`.

File: public/app/core/components/PageHeader/PageHeader.tsx [hunk 11]
```diff
@@ -71,86 +71,92 @@ const Navigation = ({ children }: { children: NavModelItem[] }) => {
...
-            {children && children.length && <Navigation>{children}</Navigation>}
-          </div>
-        </div>
-      </div>
-    );
   }
+  breadcrumbsResult.push(<span key={breadcrumbs.length + 1}> / {title}</span>);
+
+  return (
+    <h1 className="page-header__title" data-testid="page-title">
```
Comment (Clarity-89) [inline public/app/core/components/PageHeader/PageHeader.tsx:149]: Same here with the `data-testid`.

**Review by Clarity-89 — APPROVED** [review]
Looks good! Left a few comments about testing.
